### ISSUE TITLE
samtools_bam_to_cram does not store genome delta's

### ISSUE THREAD

**Issue by alpapan** [galaxyproject/tools-iuc#6523]
Problem described here https://github.com/GMOD/jbrowse-components/issues/4637

Basically, i'm not sure why it is using the `--output-fmt-option no_ref` since it requires a reference just above that line.

```
            -T "\$reffa"
            -t "\$reffai"
```

by including --output-fmt-option no_ref, we lose the differences to the genome which makes jbrowse2 visualisation not work correctly. was there a benefit using this option?

**Comment by bernt-matthias** [comment]
The problem is that the cram format would not store the reference, but only the location of the reference. In distributed compute settings (pulsar) the location of the reference will be different during the job. Hence it won't 

Maybe the new `embed_ref=0|1` output option could be of help?

**Comment by cmdcolin** [comment]
>The problem is that the cram format would not store the reference, but only the location of the reference

This problem is not unique to "hpc distributed compute". It is applies to almost anyone who uses CRAM for basically any purpose. People who use CRAM should be (I would claim: they are) accustomed to explicitly providing the reference genome via a -T argument (e.g. in the case of using samtools) anytime they use it, because the automatic lookup mechanisms are quite limited. Samtools is pretty good at throwing an error when users do not specify the reference properly or the reference is not found or the wrong reference is provided (the error may be confusing, but it occurs)

The no_ref flag "materially changes how CRAM data is stored" to not include any information about SNPs. it basically removes the concept of using "reference based compression" which is what CRAM is designed for. It could have complex downstream effects on variant analysis that users do. You can see in the jbrowse thread: there are no SNPs! That's because with no_ref, no SNPs are encoded in the file! 

Any tool that is faced with a no_ref file would have to do "heroics" (read: things it shouldn't have to do) to reconstruct the SNPs by comparing the CRAM file against the reference genome (again, the reference genome has to get looped back at some point!) so why not just encode the delta between the reference genome and the sequence data in the CRAM file itself?

**Comment by bernt-matthias** [comment]
So, what would be a solution to this problem?

**Comment by cmdcolin** [comment]
turn off using no_ref=1 by default. there is no problem that is being solved by it

**Comment by cmdcolin** [comment]
if it is not clear what i mean by "there is no problem being solved by it", i mean specifically that CRAM always requires a reference file to be retrieved at some point. CRAM decoding tools almost always allow user specified reference genomes to be added. so instead of saying "we can't know where the reference is, so we will add no_ref" the answer is "we will use the default, we don't need to add no_ref, because if the user receives any error, then they will know to add a -T argument to samtools to specify their reference genome path"

**Comment by bernt-matthias** [comment]
FYI https://github.com/galaxyproject/tools-iuc/pull/1963#discussion_r207997309

Also it seems that we use `no_ref` in many tools. 

I try to draw attention to this issue. If we fix it, we will need to fix it it quite a few places.

**Comment by alpapan** [comment]
> FYI https://github.com/galaxyproject/tools-iuc/pull/1963#discussion_r207997309
> 
> Also it seems that we use `no_ref` in many tools. 
> 
> I try to draw attention to this issue. If we fix it, we will need to fix it it quite a few places. 
> 
> 

Fingers crossed this won't lead to any GWAS paper retractions :-)

**Comment by cmdcolin** [comment]
I did a little more analysis of data files that are created using the no_ref option. I tried for example, creating a "samtools mpileup" on a file with and without no_ref added. To my surprise, the mpileup's turn out they are the same. If I delete the original reference file, running mpileup on the cram file without no_ref causes an error, but running it on the cram file with no_ref still works, and, the mpileup is still bit identical with when the reference file was still on disk.

This was interesting and surprising to me so I had to figure out how the no_ref was resolving the genome delta's (e.g. the SNPs...). Turns out samtools uses a trick thatJBrowse was not aware :  the trick is, I believe, that no_ref files are actually given an MD tag. This was surprising because MD is normally used in BAM and I was not aware of it's usage in CRAM, but with the MD tag in a CRAM file with no_ref, you CAN actually calculate the genome deltas. 

JBrowse, in this case, can add a feature (or bugfix depending on how you view it) to use the MD tag in CRAM files with no_ref to solve the bug that @alpapan initially reported to us.

I still consider it somewhat odd to use the no_ref behavior but I can walk back my hard-line stance against it. It looks like Galaxy has been using it since 2018

I don't have too many external links on this as information is sparse on the web about no_ref but i hope this helps

**Comment by cmdcolin** [comment]
I would be curious if you have any further thoughts on this @jkbonfield or any other context for no_ref behavior, but with the MD tag, jbrowse should be able to get the SNPs from no_ref cram files

**Comment by cmdcolin** [comment]
thanks so much for providing this perspective @jkbonfield, I may have overstated the importance of the MD tag in my above assessment, I'll have to do maybe even a further analysis

> "However it would be far better to use an embedded reference instead. Newer htslib even defaults to something similar to this if no external reference can be identified. "

do you have any examples of CRAM files that use these techniques?

**Comment by jkbonfield** [comment]
Not online, because ENA can't accept them yet as far as I'm aware :-(

However you can produce them easily in samtools.

```
# external reference based
samtools view -O cram -T x.ref in.bam -o out.cram

# embedded reference
samtools view -O cram,embed_ref -T x.ref in.bam -o out.cram

# embedded consensus
samtools view -O cram,embed_ref=2 in.bam -o out.cram

# no reference
samtools view -O cram,no_ref in.bam -o out.cram
```

You can verify what it's doing with `samtools cram-size out.cram` too and look for the various blocks being used and their relative sizes.  Also this shows how much space is taken up in MD and NM tags.  Also you can try `use_bzip2` or `use_lzma` (slow) for `no_ref` mode, which may dramatically reduce storage.  Gzip (deflate) just isn't that good a tool for sequence compression, and it fails dismally once we get to long-read sequences as the length of an individual sequence may be longer than the maximum gzip window.

The purpose of embedded consensus was basically to make CRAM viable for dealing with assemblies where we don't want to distribute a reference (it's just something else to go wrong).  Furthermore if we're doing something that turns CRAM into another CRAM, such as filtering, then it can automatically detect the lack of a known reference and recreate the new CRAM from scratch again.  (It's possible to extract the embedded reference and supply it as a reference file, but it's problematic and likely leads to more pitfalls about not supplying both CRAM & ref.)

Using an external reference doesn't save *that* much either.  It's basically 2-bits per covered base regardless of alignment depth.  For very shallow data that's ~700Mb off a human genome, which may be a significant portion.  For deep data it's still ~700Mb less, but that's perhaps only 5-10% of the storage costs.   Most of the gain over BAM came from collating data into columns and smarter compression codecs.

**Comment by cmdcolin** [comment]
thanks, that should be a good start!

**Comment by alpapan** [comment]
it seems that this was due to an old 2018 decision https://github.com/galaxyproject/tools-iuc/pull/1963 (according to a comment i found in iuc/samtools_view/32dc5f781059/samtools_view/samtools_view.xm)

**Comment by cmdcolin** [comment]
Update from the JBrowse side: we found a workaround, so these files will show SNPs correctly regardless of what happens to the flag here.

cram-js now diffs the verbatim bases in a no_ref CRAM against whatever reference the browser already has, and reports the differences that fall out — GMOD/cram-js#36. It's on main and goes out in the next release; the JBrowse alignments plugin tracks `^13.3.0`, so it picks the fix up with no change on that side.

**Correcting myself on the MD tag.** Earlier in this thread I said no_ref files get an MD tag and that JBrowse should be able to pull the SNPs out of it. That was wrong, and @jkbonfield's reply is the accurate version: htslib doesn't synthesise MD in no_ref mode, it only carries one through if the source file already had it. So MD was never the mechanism — and it turns out not to be needed. The bases sit verbatim in a `b` read feature, so a consumer that has its own reference can just diff the run against it. Nothing is lost in the file; the comparison simply never happened at write time.

Tracing where `b` comes from in htslib, for the record: `cram/cram_encode.c` puts it plainly — *"Non reference based encoding means storing the bases verbatim as features"*. It's written only by an encoder with no reference, i.e. `no_ref` and `samtools depad`. There's a second path that would emit `b` with a reference present, but it's gated behind `if (0 && ...)`, so with a reference htslib always writes individual `X` features.

Some numbers, since I don't think the thread has quantified what no_ref actually costs:

- **Decode speed: no difference.** samtools reads a no_ref file as fast as a reference-based one — 0.43s either way over 200,000 reads. It never computes the delta, it just hands back the stored bases.
- **Size: about 1.7x.** Those same 200,000 reads took 493 KB with no_ref against 293 KB against a reference. Less dramatic than I'd assumed — a real CRAM is mostly quality scores, and those are identical either way. (Worth reading alongside @jkbonfield's note that `use_bzip2`/`use_lzma` cut the no_ref sequence blocks substantially.)
- The cost lands on whoever computes the differences, at roughly 7ns per aligned base in our implementation. Real, but small.

So performance isn't the argument against no_ref, and I don't want the workaround to read as "this is fine now" either. The real cost is that the file stops answering "how does this read differ from the reference" on its own, and every downstream tool has to reimplement that against a reference it fetches itself. We happened to be in a position to do that; most tools aren't.

**Comment by jkbonfield** [comment]
It's worth considering htslib's embed_ref=2 too.  For aligned data, this generates a reference-coordinate consensus and embeds that, much like the normal embed_ref(=1) would do if supplying a reference while creating a CRAM (in Samtools anyway, ie `samtools view -O cram,embed_ref -T ref.fa foo.bam -o foo.cram`).

With `embed_ref=2` the external reference no longer needs to be supplied.  Coordinates still match as the CIGAR strings are used to intuit the reference coordinate system and even if every read has an insertion or a deletion it won't appear in the consensus so the CIGAR string and positions are correct.  However the impact here is that any MD and NM tags *may* need to be stored verbatim, if present, as the consensus may well be difference to the reference.  It was written mainly as a mechanism for people using embedded references that still need a copy of it if they're operating on cram.  Eg imagine `samtools view -F 0xf00 -o out.cram in.cram`.  If in.cram has an embedded reference (eg snippets of GRCh38) then the above doesn't work and we'd need to extract the reference from in.cram and resupply it to out.cram via `-T` again.  Messy and error prone.  Instead it just switches to on-the-fly consensus reference creation.

With hindsight it's probably what should have been done all along as pragmatically it's the only way a tool chain can work without requiring extra effort from the user.

Anyway, bottom line, compression ratios are much better than using no-ref as Deflate is pretty rubbish as optimising the sequence deduplication.
